src/sockaddr.rs

```diff
@@ -312,44 +313,196 @@ impl fmt::Debug for SockAddr {
     }
 }
 
-#[test]
-fn ipv4() {
-    use std::net::Ipv4Addr;
-    let std = SocketAddrV4::new(Ipv4Addr::new(1, 2, 3, 4), 9876);
-    let addr = SockAddr::from(std);
-    assert!(addr.is_ipv4());
-    assert_eq!(addr.family(), AF_INET as sa_family_t);
-    assert_eq!(addr.domain(), Domain::IPV4);
-    assert_eq!(addr.len(), size_of::<sockaddr_in>() as socklen_t);
-    assert_eq!(addr.as_socket(), Some(SocketAddr::V4(std)));
-    assert_eq!(addr.as_socket_ipv4(), Some(std));
-    assert!(addr.as_socket_ipv6().is_none());
-
-    let addr = SockAddr::from(SocketAddr::from(std));
-    assert_eq!(addr.family(), AF_INET as sa_family_t);
-    assert_eq!(addr.len(), size_of::<sockaddr_in>() as socklen_t);
-    assert_eq!(addr.as_socket(), Some(SocketAddr::V4(std)));
-    assert_eq!(addr.as_socket_ipv4(), Some(std));
-    assert!(addr.as_socket_ipv6().is_none());
+impl PartialEq for SockAddr {
+    fn eq(&self, other: &Self) -> bool {
+        unsafe {
+            let these_bytes: &[u8] = any_as_u8_slice(&self.storage, self.len as usize);
+            let those_bytes: &[u8] = any_as_u8_slice(&other.storage, other.len as usize);
+            these_bytes == those_bytes
```

Comment:
@Thomasdezeeuw what do you think about this implementation? I am casting `storage` values to a byte slices and comparing them.

The approach that you suggested [here](https://github.com/rust-lang/socket2/pull/373#discussion_r1059961032) will require some special treatment for the case when `SockAddr` is neither `AF_INET` or `AF_INET6` and I don't have a good solution for this case.

Reply:
Also what do you think about adding quickcheck test dependency to the project? It is for the property based testing.

See https://github.com/BurntSushi/quickcheck

I could use it for more comprehensive testing of `Eq` and `Hash` implementations.

Reply:
Sorry, I don't have a lot of time to review this atm, so I just took a quick look.

> Uninitialized fields are set to `0` by the `From` implementations:
> 
> ```
> let mut storage = unsafe { mem::zeroed::<sockaddr_storage>() };
> ```

That's not the only way to create a `SockAddr`, so we can't depending on this being true. We simply cannot access storage beyond `len` bytes.

> `SockAddr::new` rustdoc advises to do the same before populating `sockaddr_storage` and passing it to the constructor.
> 
> The approach that you suggested [here](https://github.com/rust-lang/socket2/pull/373#discussion_r1059961032) will require some special treatment for the case when `SockAddr` is neither `AF_INET` nor `AF_INET6` and I don't have a good solution for this case. I can panic or return `false` which would not work for other protocol families/domains.

Please don't panic in you code, it's very rarely the right thing to do.

Reply:
yep... no panic here.

> We simply cannot access storage beyond `len` bytes.
and that is exactly what `any_as_u8_slice` does, it creates a byte slice limited by `len`, see the 2nd argument:
```
let these_bytes: &[u8] = any_as_u8_slice(&self.storage, self.len as usize);
let those_bytes: &[u8] = any_as_u8_slice(&other.storage, other.len as usize);
```